{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         gene1  gene2  minResCount\n",
      "0            2     C2            3\n",
      "1            2     C3            8\n",
      "2            2    C4B            8\n",
      "3            2  C4BPA            2\n",
      "4            2     C5            8\n",
      "...        ...    ...          ...\n",
      "1322480  ZWINT  ZWINT         3828\n",
      "1322481   ZXDA   ZXDA            5\n",
      "1322482   ZXDA   ZXDC            5\n",
      "1322483   ZXDC   ZXDC            5\n",
      "1322484    ZYX    ZYX            6\n",
      "\n",
      "[1322485 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "ggLink = pd.read_csv('ggLink.csv', sep='\\t')\n",
    "print(ggLink[['gene1','gene2','minResCount']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>linkTypes</th>\n",
       "      <th>pairCount</th>\n",
       "      <th>oppCount</th>\n",
       "      <th>docCount</th>\n",
       "      <th>dbList</th>\n",
       "      <th>minResCount</th>\n",
       "      <th>snippet</th>\n",
       "      <th>context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ERVK-6</td>\n",
       "      <td>text</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>AHF1 and AHF2 were of the same amino acid sequ...</td>\n",
       "      <td>Hemorrhage (1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC3D</td>\n",
       "      <td>ppi</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>string</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC3F</td>\n",
       "      <td>ppi</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>string</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC3G</td>\n",
       "      <td>ppi</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>string</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>A2M</td>\n",
       "      <td>AKT1</td>\n",
       "      <td>text</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>The transient expression of constitutively act...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1322463</th>\n",
       "      <td>ZP1</td>\n",
       "      <td>ZP3</td>\n",
       "      <td>pwy,text,rev</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>reactome</td>\n",
       "      <td>0</td>\n",
       "      <td>Neither Zp3 transcripts nor ZP3 protein was de...</td>\n",
       "      <td>Infertility, Female (1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1322464</th>\n",
       "      <td>ZP1</td>\n",
       "      <td>ZP4</td>\n",
       "      <td>pwy,fwd</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>reactome</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1322465</th>\n",
       "      <td>ZP2</td>\n",
       "      <td>ZP3</td>\n",
       "      <td>pwy,text,fwd</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>reactome</td>\n",
       "      <td>0</td>\n",
       "      <td>Neither Zp3 transcripts nor ZP3 protein was de...</td>\n",
       "      <td>Infertility, Female (1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1322466</th>\n",
       "      <td>ZP2</td>\n",
       "      <td>ZP4</td>\n",
       "      <td>pwy,fwd</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>reactome</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1322468</th>\n",
       "      <td>ZP3</td>\n",
       "      <td>ZP4</td>\n",
       "      <td>pwy,fwd</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>reactome</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>514296 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene1     gene2     linkTypes  pairCount  oppCount  docCount  \\\n",
       "15       A1BG    ERVK-6          text          0         1         1   \n",
       "25       A1CF  APOBEC3D           ppi          0         0         0   \n",
       "26       A1CF  APOBEC3F           ppi          0         0         0   \n",
       "27       A1CF  APOBEC3G           ppi          0         0         0   \n",
       "38        A2M      AKT1          text          1         0         1   \n",
       "...       ...       ...           ...        ...       ...       ...   \n",
       "1322463   ZP1       ZP3  pwy,text,rev          0         1         1   \n",
       "1322464   ZP1       ZP4       pwy,fwd          0         0         0   \n",
       "1322465   ZP2       ZP3  pwy,text,fwd          0         1         1   \n",
       "1322466   ZP2       ZP4       pwy,fwd          0         0         0   \n",
       "1322468   ZP3       ZP4       pwy,fwd          0         0         0   \n",
       "\n",
       "           dbList  minResCount  \\\n",
       "15            NaN            0   \n",
       "25         string            0   \n",
       "26         string            0   \n",
       "27         string            0   \n",
       "38            NaN            0   \n",
       "...           ...          ...   \n",
       "1322463  reactome            0   \n",
       "1322464  reactome            0   \n",
       "1322465  reactome            0   \n",
       "1322466  reactome            0   \n",
       "1322468  reactome            0   \n",
       "\n",
       "                                                   snippet  \\\n",
       "15       AHF1 and AHF2 were of the same amino acid sequ...   \n",
       "25                                                     NaN   \n",
       "26                                                     NaN   \n",
       "27                                                     NaN   \n",
       "38       The transient expression of constitutively act...   \n",
       "...                                                    ...   \n",
       "1322463  Neither Zp3 transcripts nor ZP3 protein was de...   \n",
       "1322464                                                NaN   \n",
       "1322465  Neither Zp3 transcripts nor ZP3 protein was de...   \n",
       "1322466                                                NaN   \n",
       "1322468                                                NaN   \n",
       "\n",
       "                         context  \n",
       "15                Hemorrhage (1)  \n",
       "25                           NaN  \n",
       "26                           NaN  \n",
       "27                           NaN  \n",
       "38                           NaN  \n",
       "...                          ...  \n",
       "1322463  Infertility, Female (1)  \n",
       "1322464                      NaN  \n",
       "1322465  Infertility, Female (1)  \n",
       "1322466                      NaN  \n",
       "1322468                      NaN  \n",
       "\n",
       "[514296 rows x 10 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ggLink[ggLink['minResCount']==0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_version(x):\n",
    "    if '.' in x:\n",
    "        return x.split('.')[0]\n",
    "    return x\n",
    "\n",
    "gtf_file_path = '/work/h2020deciderficarra_shared/TCGA/OV/project_n16_data/gencode.v47.annotation.gtf'\n",
    "gtf = pd.read_csv(gtf_file_path, sep=\"\\t\", header=None, comment='#')\n",
    "gtf.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "\n",
    "parameters = ['gene_id', 'gene_type', 'gene_name']\n",
    "for p in parameters:\n",
    "    gtf[p] = gtf['attribute'].apply(lambda x: re.findall(rf'{p} \"([^\"]*)\"', x)[0] if rf'{p} \"' in x else np.nan)\n",
    "\n",
    "gtf.drop(['attribute','source','score','strand','frame'], axis=1, inplace=True)\n",
    "\n",
    "gtf['gene_id'] = gtf['gene_id'].apply(remove_version)\n",
    "\n",
    "gtf_pc = gtf[gtf['gene_type'] == 'protein_coding']\n",
    "\n",
    "# Protein coding set\n",
    "pc_set = set(gtf_pc['gene_id'].to_list())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_pc_no_dup = gtf_pc.drop_duplicates(subset=['gene_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "si\n"
     ]
    }
   ],
   "source": [
    "gene_name1 = gtf_pc_no_dup[gtf_pc_no_dup['gene_id'] == 'ENSG00000188976']['gene_name'].values[0]\n",
    "gene_name2 = gtf_pc_no_dup[gtf_pc_no_dup['gene_id'] == 'ENSG00000225921']['gene_name'].values[0]\n",
    "\n",
    "if not ggLink[(ggLink['gene1'] == gene_name1) & (ggLink['gene2'] == gene_name2)].empty:\n",
    "    print('si')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOL7\n"
     ]
    }
   ],
   "source": [
    "df_gene = gtf_pc_no_dup[gtf_pc_no_dup['gene_id'] == 'ENSG00000225921']['gene_name']\n",
    "if not df_gene.empty:\n",
    "    gene_name2 = df_gene.values[0]\n",
    "else:\n",
    "    gene_name2 = ''\n",
    "print(gene_name2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/homes/dlupo/.local/lib/python3.9/site-packages/torch/__config__.py:10: UserWarning: CUDA initialization: The NVIDIA driver on your system is too old (found version 11040). Please update your GPU driver by downloading and installing a new version from the URL: http://www.nvidia.com/Download/index.aspx Alternatively, go to: https://pytorch.org to install a PyTorch version that has been compiled with your version of the CUDA driver. (Triggered internally at /pytorch/c10/cuda/CUDAFunctions.cpp:109.)\n",
      "  return torch._C._show_config()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "import torch\n",
    "import json\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "from torch_geometric.utils.convert import to_networkx, from_networkx\n",
    "\n",
    "PATH_FOLDER_COPY_NUMBER = \"/work/h2020deciderficarra_shared/TCGA/OV/project_n16_data/CopyNumber\"\n",
    "PATH_CASE_ID_STRUCTURE = \"/homes/dlupo/Progetto_BioInformatics/AI_for_Bioinformatics_Project/Preprocessing/Final/case_id_and_structure.json\"\n",
    "PATH_GENE_ID_PROTEIN_CODING = \"/homes/dlupo/Progetto_BioInformatics/AI_for_Bioinformatics_Project/Preprocessing/Final/gene_id_protein_coding.json\"\n",
    "\n",
    "with open(PATH_CASE_ID_STRUCTURE, 'r') as file:\n",
    "    file_parsed = json.load(file)\n",
    "\n",
    "copy_number_folder_list = []\n",
    "os_list = []\n",
    "for key in file_parsed.keys():\n",
    "    copy_number_folder_list.append(file_parsed[key][\"files\"][\"copy_number\"])\n",
    "    os_list.append(file_parsed[key]['os'])\n",
    "print()\n",
    "gene_id_list = []\n",
    "with open(PATH_GENE_ID_PROTEIN_CODING) as json_file:\n",
    "   gene_id_list = json.load(json_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_version(x):\n",
    "    if '.' in x:\n",
    "        return x.split('.')[0]\n",
    "    return x\n",
    "\n",
    "list_df_CNV = []\n",
    "for root, dirs, files in os.walk(PATH_FOLDER_COPY_NUMBER):\n",
    "    for dir in dirs:\n",
    "        for root, dirs, files in os.walk(PATH_FOLDER_COPY_NUMBER + \"/\" + dir):\n",
    "            for file in files:\n",
    "                if file in copy_number_folder_list:\n",
    "                    parsed_file = pd.read_csv(PATH_FOLDER_COPY_NUMBER + \"/\" + dir + \"/\" + file, sep='\\t')\n",
    "                    parsed_file['gene_id'] = parsed_file['gene_id'].apply(remove_version)\n",
    "                    parsed_file = parsed_file[parsed_file['gene_id'].isin(gene_id_list)]\n",
    "                    list_df_CNV.append(parsed_file)\n",
    "\n",
    "list_df_CNV_filled = []\n",
    "for i in range(len(list_df_CNV)):\n",
    "    list_df_CNV_filled.append(list_df_CNV[i].fillna(0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mSi Ã¨ verificato un arresto anomalo del Kernel durante l'esecuzione del codice nella cella attiva o in una cella precedente. \n",
      "\u001b[1;31mEsaminare il codice nelle celle per identificare una possibile causa dell'errore. \n",
      "\u001b[1;31mPer altre informazioni, fare clic<a href='https://aka.ms/vscodeJupyterKernelCrash'>qui</a>. \n",
      "\u001b[1;31mPer ulteriori dettagli, visualizzare Jupyter <a href='command:jupyter.viewOutput'>log</a>."
     ]
    }
   ],
   "source": [
    "edge = [[],[]]\n",
    "# nodes = len(list_df_CNV_filled[0])\n",
    "nodes = 300\n",
    "df = list_df_CNV_filled[0]\n",
    "# Aggiunta delle connessioni in base alla sovrapposizione\n",
    "for f_1_index in range(nodes):\n",
    "    for f_2_index in range(f_1_index + 1, nodes):\n",
    "        row1 = df.iloc[f_1_index]\n",
    "        row2 = df.iloc[f_2_index]\n",
    "\n",
    "        df_gene = gtf_pc_no_dup[gtf_pc_no_dup['gene_id'] == row1['gene_id']]['gene_name']\n",
    "        if not df_gene.empty:\n",
    "            gene_name1 = df_gene.values[0]\n",
    "        else:\n",
    "            gene_name1 = ''\n",
    "\n",
    "        df_gene = gtf_pc_no_dup[gtf_pc_no_dup['gene_id'] == row2['gene_id']]['gene_name']\n",
    "        if not df_gene.empty:\n",
    "            gene_name2 = df_gene.values[0]\n",
    "        else:\n",
    "            gene_name2 = ''\n",
    "\n",
    "        if not ggLink[(ggLink['gene1'] == gene_name1) & (ggLink['gene2'] == gene_name2)].empty:\n",
    "            edge[0].append(f_1_index)\n",
    "            edge[0].append(f_2_index)\n",
    "            edge[1].append(f_2_index)\n",
    "            edge[1].append(f_1_index)\n",
    "            \n",
    "print(len(edges[0]))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
